### PR TITLE
Add functionality to track LibraryVersion dependencies

### LINKED ISSUE #1402: Save dependencies for each LibraryVersion

### PR DESCRIPTION
**Pull request by brianjp93**
This pr adds:
- A new M2M field on LibraryVersion to track dependencies between libraries
- A GitHub workflow that generates dependency reports using boostdep
- Management command to update LibraryVersion dependencies from GH artifacts
- A daily task to check for the boostdep artifact and update dependencies
- Extended GithubAPIClient to fetch artifacts from GitHub Actions
- Tests for the new dependency parsing functionality
- fixes #1402

The workflow analyzes each Boost release tag of the form "boost-x.x.0" to generate a dependency report which is then used to populate the dependencies M2M between LibraryVersion and Library models.

### CODE REVIEW

**Comment by brianjp93** [comment]
Github action runs on push to master or manually.

**Comment by sdarwin** [comment]
Near the beginning of the github actions it's checking out the whole repository and then discarding it.  Here are more efficient alternatives.
- Simply download the file in a one-liner without checkout out the repo.  (if this works it seems preferable).

`curl -OL https://github.com/${GITHUB_REPOSITORY}/blob/${GITHUB_SHA}/.github/workflows/scripts/get_deps.sh?raw=true`

or

- Switch to sparse mode:  https://github.com/actions/checkout

```
# Fetch only a single file
- uses: actions/checkout@v4
  with:
    sparse-checkout: |
      .github/workflows/scripts/get_deps.sh
    sparse-checkout-cone-mode: false
```

**Comment by sdarwin** [comment]
the PR is ok with me, maybe Greg can review the django code.

File: libraries/models.py [hunk 1]
```diff
@@ -378,6 +378,12 @@ class LibraryVersion(models.Model):
     deletions = models.IntegerField(default=0)
     files_changed = models.IntegerField(default=0)
     cpp_standard_minimum = models.CharField(max_length=50, blank=True, null=True)
+    dependencies = models.ManyToManyField(
+        "Library",
```
Comment (GregKaleka) [inline libraries/models.py:382]: Above, the FK to library uses `libraries.Library`. Doesn't really matter, but we should probably be consistent.

File: libraries/tests/fixtures.py [hunk 2]
```diff
@@ -226,3 +226,19 @@ def github_library():
...
+    dependency_report.yaml action
+    """
+    return """Dependencies for version boost-1.33.0
+Dependencies for version boost-1.34.0
+Dependencies for version boost-1.35.0
+algorithm -> concept_check config detail logic numeric~conversion
+Dependencies for version boost-1.85.0
+algorithm -> array assert bind concept_check config core
+numeric~conversion -> array
+callable_traits ->"""
```
Comment (GregKaleka) [inline libraries/tests/fixtures.py:244]: You can use `textwrap.dedent` to make this a bit nicer.

```suggestion
    return dedent(
        """\
        Dependencies for version boost-1.33.0
        Dependencies for version boost-1.34.0
        Dependencies for version boost-1.35.0
        algorithm -> concept_check config detail logic numeric~conversion
        Dependencies for version boost-1.85.0
        algorithm -> array assert bind concept_check config core
        numeric~conversion -> array
        callable_traits ->"""
    )
```

File: libraries/tests/test_github.py [hunk 3]
```diff
@@ -326,3 +326,63 @@ def test_update_prs_existing(
...
+    assert lv.dependencies.all().count() == 6
+    # callable traits is in the file but has no dependencies
+    lv = LibraryVersion.objects.get(
+        library__key="callable_traits", version__name="boost-1.85.0"
+    )
+    assert lv.dependencies.all().count() == 0
+    lv = LibraryVersion.objects.get(
+        library__key="numeric/conversion", version__name="boost-1.85.0"
+    )
+    assert lv.dependencies.all().count() == 1
```
Comment (GregKaleka) [inline libraries/tests/test_github.py:388]:
```suggestion
    assert lv.dependencies.count() == 5
    lv = LibraryVersion.objects.get(
        library__key="algorithm", version__name="boost-1.85.0"
    )
    assert lv.dependencies.count() == 6
    # callable traits is in the file but has no dependencies
    lv = LibraryVersion.objects.get(
        library__key="callable_traits", version__name="boost-1.85.0"
    )
    assert lv.dependencies.count() == 0
    lv = LibraryVersion.objects.get(
        library__key="numeric/conversion", version__name="boost-1.85.0"
    )
    assert lv.dependencies.count() == 1
```

File: libraries/utils.py [hunk 4]
```diff
@@ -210,3 +210,62 @@ def legacy_path_transform(content_path):
     if content_path and content_path.startswith(LEGACY_LATEST_RELEASE_URL_PATH_STR):
         content_path = re.sub(r"([a-zA-Z0-9\.]+)/(\S+)", r"latest/\2", content_path)
     return content_path
+
+
+def parse_boostdep_artifact(content: str):
```
Comment (GregKaleka) [inline libraries/utils.py:215]: Can you add a docstring?

File: libraries/github.py [hunk 5]
```diff
@@ -587,3 +586,44 @@ def update_commit_author_github_data(self, obj=None, email=None, overwrite=False
                 if gh_author["html_url"]:
                     author.github_profile_url = gh_author["html_url"]
                 author.save(update_fields=["avatar_url", "github_profile_url"])
+
+    def fetch_most_recent_boost_dep_artifact_content(self, owner=None):
```
[the remark this answers is not in the document]
Comment (brianjp93) [reply libraries/github.py:590]: It's not optional to github, but `get_artifacts` falls back to `self.owner` if the kwarg is not supplied.  I can set it to an empty string by default, though.

File: libraries/github.py [hunk 6]
```diff
@@ -587,3 +586,44 @@ def update_commit_author_github_data(self, obj=None, email=None, overwrite=False
...
+        artifacts = self.client.get_artifacts(
+            owner=owner,
+            repo="website-v2",
+            name="boost-dep-artifact",
+        )
+        if not artifacts or not artifacts.get("artifacts", None):
+            logger.warning("No artifacts found.")
+            return
+        artifact = artifacts["artifacts"][0]
```
[the remark this answers is not in the document]
Comment (brianjp93) [reply libraries/github.py:599]: Yes, the most recent one.  I'll add a note.

File: libraries/management/commands/update_library_version_dependencies.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+@click.command()
+@click.option("--token", is_flag=False, help="Github API token")
+@click.option(
+    "--owner",
+    is_flag=False,
+    help="The repo owner of the artifact to be parsed for dependencies.",
+)
+@click.option("--clean", is_flag=True, help="Remove dependencies before adding them.")
+def command(token=None, owner=None, clean=False):
```
[the remark this answers is not in the document]
Comment (brianjp93) [reply libraries/management/commands/update_library_version_dependencies.py:14]: Yes, it falls back to that in LibraryUpdater

File: libraries/tests/fixtures.py (hunk 2 above)
Comment (brianjp93) [reply libraries/tests/fixtures.py:244]: Never used that before.  Good to know.

File: libraries/utils.py (hunk 4 above)
Comment (brianjp93) [reply libraries/utils.py:215]: I think I addressed all of the concerns, let me know if there is anything left.

**Pull request merged**